src/analysis/normed_space/real_inner_product.lean

```diff
@@ -0,0 +1,448 @@
+by { repeat {rw [← pow_two]}, exact norm_add_pow_two }
+
+/-- Expand the square -/
+lemma norm_sub_pow_two {x y : α} : ∥x - y∥^2 = ∥x∥^2 - 2 * inner x y + ∥y∥^2 :=
+by { repeat {rw [pow_two, ← inner_self_eq_norm_square]}, exact inner_sub_sub_self }
+
+/-- Same lemma as above but in a different form -/
+lemma norm_sub_mul_self {x y : α} : ∥x - y∥ * ∥x - y∥ = ∥x∥ * ∥x∥ - 2 * inner x y + ∥y∥ * ∥y∥ :=
+by { repeat {rw [← pow_two]}, exact norm_sub_pow_two }
+
+/-- Cauchy–Schwarz inequality with norm -/
+lemma abs_inner_le_norm (x y : α) : abs (inner x y) ≤ ∥x∥ * ∥y∥ :=
+nonneg_le_nonneg_of_squares_le (mul_nonneg (sqrt_nonneg _) (sqrt_nonneg _))
+begin
+  rw abs_mul_abs_self,
+  have : ∥x∥ * ∥y∥ * (∥x∥ * ∥y∥) = inner x x * inner y y,
+    simp only [inner_self_eq_norm_square], ring,
+  rw this,
+  exact inner_mul_inner_self_le _ _
+end
+
+lemma parallelogram_law_with_norm {x y : α} :
+  ∥x + y∥ * ∥x + y∥ + ∥x - y∥ * ∥x - y∥ = 2 * (∥x∥ * ∥x∥ + ∥y∥ * ∥y∥) :=
+by { simp only [(inner_self_eq_norm_square _).symm], exact parallelogram_law }
+
+/-- An inner product space forms a normed group w.r.t. its associated norm. -/
+instance inner_product_space_is_normed_group : normed_group α :=
+normed_group.of_core α
+{ norm_eq_zero_iff := assume x, iff.intro
+    (λ h : sqrt (inner x x) = 0, (inner_self_eq_zero x).1 $ (sqrt_eq_zero inner_self_nonneg).1 h )
+    (by {rintro rfl, show sqrt (inner (0:α) 0) = 0, simp }),
+  triangle := assume x y,
+  begin
+    have := calc
+      ∥x + y∥ * ∥x + y∥ = inner (x + y) (x + y) : (inner_self_eq_norm_square _).symm
+      ... = inner x x + 2 * inner x y + inner y y : inner_add_add_self
+      ... ≤ inner x x + 2 * (∥x∥ * ∥y∥) + inner y y :
+        by linarith [abs_inner_le_norm x y, le_abs_self (inner x y)]
+      ... = (∥x∥ + ∥y∥) * (∥x∥ + ∥y∥) : by { simp only [inner_self_eq_norm_square], ring },
+    exact nonneg_le_nonneg_of_squares_le (add_nonneg (sqrt_nonneg _) (sqrt_nonneg _)) this
+  end,
+  norm_neg := λx, show sqrt (inner (-x) (-x)) = sqrt (inner x x), by simp }
+
+/-- An inner product space forms a normed space over reals w.r.t. its associated norm. -/
+instance inner_product_space_is_normed_space : normed_space ℝ α :=
+{ norm_smul := assume r x,
+  begin
+    rw [norm_eq_sqrt_inner, sqrt_eq_iff_mul_self_eq,
+        inner_smul_left, inner_smul_right, inner_self_eq_norm_square],
+    exact calc
+      abs(r) * ∥x∥ * (abs(r) * ∥x∥) = (abs(r) * abs(r)) * (∥x∥ * ∥x∥) : by ring
+      ... = r * (r * (∥x∥ * ∥x∥)) : by { rw abs_mul_abs_self, ring },
+    exact inner_self_nonneg,
+    exact mul_nonneg (abs_nonneg _) (sqrt_nonneg _)
+  end }
+
+end norm
+
+section orthogonal
+
+open filter
+
+/--
+Existence of minimizers
+Let `u` be a point in an inner product space, and let `K` be a nonempty complete convex subset.
+Then there exists a unique `v` in `K` that minimizes the distance `∥u - v∥` to `u`.
+ -/
+theorem exists_norm_eq_infi_of_complete_convex {K : set α} (ne : nonempty K) (h₁ : is_complete K)
```

Comment:
Wow, this is a surprising amount of work.

Reply:
Thanks!